### ISSUE TITLE
ConvTranspose is not accepting negative padsize

### ISSUE THREAD

**Issue by stexa** [sonos/tract#991]
Hey! First of all thanks for your amazing progress with tract!
I have an ONNX model that is using ConvTranspose (conv2d) with negative padsize (I guess to crop).
The attributes of the operation look like that:
- auto_pad: NOTSET
- dilitations: 1, 1
- kernel_shape: 1, 3
- pads: 0, 0, 0, -1
- strides: 1, 2

Tract is refusing to load the model with `attribute 'pads': expected list of non-negative ints`.
Is there a possibility for tract to understand this?

The model was automatically translated by tf2onnx and is working fine with onnx-runtime.

**Comment by kali** [comment]
Hey, thanks for the nice words.

Well, *sigh*, this is a new one. The creativity of model designers in finding new ways of abusing the spec will never cease to amaze me.

Can you give me the model file and an input spec ? I'll see how I can workaround this one.

**Comment by stexa** [comment]
Here is an example model with pad size of -1 and the python code which is generating this weird operation.

`
import tensorflow as tf
import os
import subprocess

# we want to go (128, 32, 1) -> (128, 16, 1) -> (128, 32, 1) using conv and tconv layers

# define keras model with conv + pad and transposed conv + crop layers
def get_model():
    model = tf.keras.models.Sequential()
    model.add(tf.keras.layers.InputLayer(input_shape=(128, 32, 1)))
    model.add(tf.keras.layers.ZeroPadding2D(padding=(0, 1)))
    model.add(
        tf.keras.layers.Conv2D(
            filters=1, kernel_size=(1, 3), strides=(1, 2), padding="valid"
        )
    )
    model.add(
        tf.keras.layers.Conv2DTranspose(
            filters=1,
            kernel_size=(1, 3),
            strides=(1, 2),
            padding="valid",
            output_padding=(0, 1),
        )
    )
    model.add(tf.keras.layers.Cropping2D(cropping=((0, 0), (1, 1))))
    return model


# get a model
model = get_model()
x = tf.random.normal((1, 128, 32, 1))
y_hat = model.predict(x)
print(y_hat.shape)  # [1, 128, 32, 1]

# save as tf saved model
tf.saved_model.save(model, "saved_conv_tconv_model")

# export to onnx format
subprocess.run(
    [
        "python",
        "-m",
        "tf2onnx.convert",
        "--saved-model",
        "saved_conv_tconv_model",
        "--output",
        os.path.join("saved_conv_tconv_model", "model.onnx"),
        "--opset",
        "17",
    ]
)
# saved_conv_tconv_model/model.onnx will have a ConvTranspose op with pad: (0, 0, 0. -1)
`

[model.zip](https://github.com/sonos/tract/files/10900696/model.zip)

**Comment by kali** [comment]
Thanks. Repro-ed. I'll need to think a bit about to workaround this, it's a bit of a mess.

**Comment by stexa** [comment]
Thank you so much!

I solved the problem now by changing something in the model.

But now something in my model hits:

``` thread 'main' panicked at 'not yet implemented', [...]/tract-hir-0.19.7/src/ops/array/reshape.rs:131:17```

Do you have an idea what could lead to this?
Unfortunately I can't share the model right now, but if I could find out which node is leading to this I could try to make another small one.

**Comment by kali** [comment]
Apologies for the terrible experience tract is giving you right now. You're hitting the todo!() here. Oops.

https://github.com/sonos/tract/blob/main/hir/src/ops/array/reshape.rs#L131

So yes, short of giving me a model, it would help if you could isolate the node that triggered the issue, along with its input shapes and attributes. And I can confirm it's highly likely you should start by looking at the Reshape nodes.

**Comment by stexa** [comment]
My current guess is, that the reshape function is not properly working because the shapes could not get fully inferred by onnx. In Netron at some connections the shapes are unknown. Do you know if that is something that could lead to the error above?

**Comment by stexa** [comment]
Yep there I get an error at the declutter stage.

```
Error at stage declutter

    Caused by:
        0: running pass PushSliceUp
        1: Trying to substitute a 1,2,32,F32 by 1,4,32,F32.
```
I didn't specify the inputs because I didn't know how to specify multiple inputs, do you have an example for that?

**Comment by kali** [comment]
Just multiple  `-i` should do the trick.

**Comment by stexa** [comment]
yep multiple -i are working.
Then I get this error in smallvec:

```
thread 'main' panicked at 'assertion failed: index < len', [..]/.cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-1.10.0/src/lib.rs:1065:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: tract_hir::ops::array::squeeze::Squeeze::output_shape
   4: <tract_onnx::ops::array::squeeze::Squeeze13 as tract_hir::ops::expandable::Expansion>::rules::{{closure}}
   5: <tract_hir::infer::rules::solver::Given2Rule<a,b> as tract_hir::infer::rules::solver::Rule>::apply
   6: tract_hir::infer::rules::solver::Solver::infer_facts
   7: tract_hir::infer::rules::<impl tract_hir::infer::ops::InferenceOp for O>::infer_facts
   8: tract_hir::infer::ops::InferenceOp::infer
   9: tract_hir::infer::analyser::Analyser<M>::analyse_obstinate
  10: <tract_core::model::graph::Graph<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn tract_hir::infer::ops::InferenceOp>> as tract_hir::infer::model::InferenceModelExt>::analyse
  11: tract::params::Parameters::pipeline
  12: tract::params::Parameters::from_clap
  13: tract::main
```

**Comment by stexa** [comment]
probably it is a problem, that the order of inputs is not equal to the order of outputs anymore in the onnx model...

**Comment by kali** [comment]
OK, so it happens during analyse, not in typing as I though. Can you try running with `-vvv` ? You should get a very verbose log, but some trace-level messages near the end should give us some hints about how this squeeze operator is constrained.

Also `--onnx-ignore-output-shapes` is often useful when you want to override the inputs of an ONNX model without fixing the outputs.

**Comment by stexa** [comment]
Okay I am getting closer.
The error in the analysis stage was coming because I was giving the inputs in a different order as they were in the model (the onnx converter mixed them a bit around).
Now I am back at the error in the declutter stage, where it tries to substitute a 1,2,32,F32 by 1,4,32,F32.
Ignoring the output shapes doesn't help either.  But as a next step I will try to generate a model with just this stage to send you something as well. Thanks for the amazing support!

```
    Caused by:
        0: Error at stage declutter
        1: running pass PushSliceUp
        2: Trying to substitute a 1,2,32,F32 by 1,4,32,F32.
           ModelPatch { context: ["push slice up"], dont_apply_twice: None, model: Graph { nodes: [Node { id: 0, name: "incoming-28/0", inputs: [], op: TypedSource { fact: 2,385,1,F32 }, outputs: [2,385,1,F32 >1/0 >2/0] }, Node { id: 1, name: "Tensordot_MatMul.split-over-1.0..1", inputs: [0/0>], op: MatMulUnary { a: 385,32,1,F32 0.5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0..., axes: MatMulAxes { a_m: 1, a_k: 0, b_k: 1, b_n: 0, c_m: 2, c_n: 1 } }, outputs: [1,2,32,F32 >3/0] }, Node { id: 2, name: "Tensordot_MatMul.split-over-1.1..2", inputs: [0/0>], op: MatMulUnary { a: 385,32,1,F32 0.5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0..., axes: MatMulAxes { a_m: 1, a_k: 0, b_k: 1, b_n: 0, c_m: 2, c_n: 1 } }, outputs: [1,2,32,F32 >3/1] }, Node { id: 3, name: "Tensordot_MatMul.concat-1", inputs: [1/0>, 2/0>], op: TypedConcat { axis: 1 }, outputs: [1,4,32,F32 ] }], inputs: [0/0>], outputs: [], outlet_labels: {}, properties: {}, symbol_table:  }, inputs: {}, incoming: {0/0>: 28/0>}, shunt_outlet_by: {}, obliterate: [] } 
```

**Comment by kali** [comment]
Nothing obvious jumped to my face here. Hopefully, from where you are, you can reproduce it with something you can share.

May one or two more command line tips: --declutter-step 12 will stop after the 12th patch in decluttering, and dump the network. With this, you may be able to dichotomize around the step and see the network right before the decluttering fails.

**Comment by mathieupoumeyrolsonos** [comment]
Reading again ONNX specification for ConvTranspose, It specifies that `pads` should be positive:

https://github.com/onnx/onnx/blob/main/docs/Operators.md#convtranspose
"Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0."

So the network is invalid. I think it is a bug in tf2onnx, it may be triggered by the crop following the deconv, I'm not sure.

In any case, I think I will make this a non-fix. Plus it's not critical as you have found a workaround. What to you think ?

**Comment by stexa** [comment]
Yep I will close the ticket! Thank you again!
